[DCJ-507] Stairway with work queue enabled should be context-aware

okotsopoulos opened the pull request:
__Jira ticket__: https://broadworkbench.atlassian.net/browse/DCJ-507

## Addresses

Context-aware Stairway previously assumed that the calling thread would have mapped diagnostic context (MDC) to persist to the child threads spawned for flight execution.

This was not the case when Stairway is configured with a work queue enabled (e.g. GCP Pub-Sub).  Here, flight information was written to the pub-sub topic without the calling thread's context, so when a message was processed and deserialized any logs emitted from that flight would be missing context set by the original calling thread (e.g. request ID).

Where we've felt this impact in TDR:
- Child flights submitted within the execution of a parent flight are missing the calling thread's context in its logs.
- While this is an uncommon pattern for flight execution, it is used when facilitating combined data ingests to datasets: the parent flight ingests tabular data, and child flights are spawned for each file being ingested.
- Much of our attention has been on debugging combined data ingests to Azure datasets: the inability to filter for all connected parent and child flight logs in GCP log console by request ID (`jsonPayload.requestId`) is slowing down investigative progress. 

## Summary of changes

- Added new helper method `MdcUtils.callWithContext` to call a callable and return its result with MDC's context map temporarily overwritten.
- Expanded `QueueMessageReady` to store calling thread context on construction.
- `QueueMessageReady.process()` sets the MDC using its stored calling thread context, which then makes the flight context-aware.

## Testing Strategy

- Added unit tests for `MdcUtils.callWithContext`.
- Expanded unit tests for `QueueMessageReady` with special attention to MDC behavior and propagation in serialization, deserialization, and processing.

~~Note: Sonarcloud coverage stats are incorrectly reporting 0% coverage.  They look to have never worked.  I do believe I've covered all of my changes in unit tests.~~ ETA: thanks @pshapiro4broad for fixing the Sonar scans, I pulled the fix into this branch and we are now seeing coverage stats.

okotsopoulos committed:
[DCJ-507] Flights queued in GCP Pub-Sub include calling thread context

Context-aware Stairway previously assumed that the calling thread would have mapped diagnostic context (MDC) to persist to the child threads spawned for flight execution.

This was not the case when Stairway is configured with a work queue enabled (e.g. GCP Pub-Sub).  Here, flight information was written to the pub-sub topic without the calling thread's context, so when a message was processed and deserialized any logs emitted from that flight would be missing context set by the original calling thread (e.g. request ID).

- Expanded QueueMessageReady to store calling thread context on construction
- QueueMessageReady.process sets the MDC using its stored calling thread context, which then makes the flight context-aware
- Expanded unit tests

okotsopoulos committed:
Sonar: JUnit 5 test classes, methods should have default visibility

rushtong reviewed (approved):
This looks reasonable to me 👍🏽

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/queue/QueueMessageReady.java:
```diff
@@ -59,4 +70,12 @@ public String getFlightId() {
     this.flightId = flightId;
   }
+
+  public Map<String, String> getCallingThreadContext() {
```
It would be nice to not have to add public APIs that are only used in tests.

For `getCallingThreadContext()`, could the test instead assert on `MDC.getCopyOfContextMap()`?

For `setCallingThreadContext()`, could the test instead call `MDC.setContextMap()` before constructing `QueueMessageReady`?

okotsopoulos replied:
Good points.

> For getCallingThreadContext(), could the test instead assert on MDC.getCopyOfContextMap()?

No, because the MDC written to the `QueueMessageReady` object on its construction can differ from the current value of the MDC.  I can remove the `public` modifier and annotate as visible for testing, though.

> For setCallingThreadContext(), could the test instead call MDC.setContextMap() before constructing QueueMessageReady?

Yes it could, and then I think I could remove this method.

okotsopoulos replied:
Update: I was able to remove the setter in full, but I did need to leave the getter as public so that serialization and deserialization would work.  This was caught by a unit test failure when I removed the public modifier, so that's nice :)

Other preexisting setters can likely be removed as well, but I'll designate that work as OOS.

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/queue/QueueMessageReady.java:
```diff
@@ -37,11 +46,13 @@ public boolean process(StairwayImpl stairwayImpl) throws InterruptedException {
               + stairwayImpl.getStairwayName()
               + (resumed ? " resumed flight: " : " did not find flight to resume: ")
               + flightId);
-      return true;
```
Was this change necessary? If it works either way, I think returning a boolean directly is clearer than creating state which must be returned at the end.

okotsopoulos replied:
Because I added a `finally` clause to restore the thread's context to its original state, I do think this was a necessary change.

okotsopoulos replied:
Ah, but it's possible that this won't be necessary if I make a change like you described here -- https://github.com/DataBiosphere/stairway/pull/152/files/93f139fd72b2e1d0b82000676a5a163b8130236e#r1676240309

I'll take a closer look.

pshapiro4broad replied:
> Because I added a finally clause to restore the thread's context to its original state, I do think this was a necessary change.

Even if you `return` inside a `try`, the `finally` will still run.

okotsopoulos replied:
I reverted to the old return strategy.

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/queue/QueueMessageReady.java:
```diff
@@ -16,18 +19,24 @@ class QueueMessageReady extends QueueMessage {
   @Override
   public boolean process(StairwayImpl stairwayImpl) throws InterruptedException {
+    boolean processed = false;
+    // Save the initial thread context so that it can be restored
+    Map<String, String> initialContext = MDC.getCopyOfContextMap();
     try {
+      MdcUtils.overwriteContext(callingThreadContext);
```
Could this be done inside `McdUtils`? Something like
```
    MdcUtils.callWithContext(callingThreadContext, () -> {
        ...
        stairwayImpl.resume(flightId)
        ...
    });
```

I think it would be better to expose less of the inner workings of `MdcUtils` if possible. And there may be other cases where we want to overwrite the context temporarily like this.

okotsopoulos replied:
Thanks for the suggestion. I added new public utility method `MdcUtils.callWithContext` and removed the `public` modifier on `MdcUtils.overwriteContext`.  Any possible refactoring of other code to use this method I'll call out of scope.

pshapiro4broad commented on stairway/src/test/java/bio/terra/stairway/queue/QueueMessageTest.java:
```diff
@@ -1,32 +1,106 @@
+
+  @BeforeEach
+  void beforeEach() {
+    MDC.clear();
```
Would it be better to use a static mock instead of the actual MDC API? We don't use mocks yet for MDC, so it may be too big a change for this PR, but it's something to consider if we'd like to isolate our tests from the MDC implementation, and might make it easier to check if our code is doing what we expect.

okotsopoulos replied:
Hmm, yeah this feels like too big a change for this PR.  But you make good points about the benefits of test isolation.

pshapiro4broad reviewed (approved):
Looks OK, although I'd prefer making fewer APIs public for tests if possible

okotsopoulos committed:
Merge branch 'develop' into okotsopo-DCJ-507-pubsub-mdc

okotsopoulos committed:
PR feedback

MdcUtils:
- Added public utility method callWithContext for running and returning a callable with MDC's context map temporarily overwritten
- Remove public modifier on overwriteContext in favor of the above

QueueMessageReady:
- Use MdcUtils.callWithContext to simplify process definition
- Revert process return strategy to return booleans directly rather than setting a boolean to return later
- Remove unnecessary callingThreadContext setter (left existing setters untouched even though they can likely be removed: OOS)
- Left public callingThreadContext getter unchanged: it is required to be public for serde operations

Added unit test coverage for all changes.

okotsopoulos committed:
Remove debugging print statement errantly checked in

fboulnois reviewed (approved):
looks good 👍
